[LBSE] Fix svg/clip-path/clip-path-on-overflowing.html

rwlbuis opened the pull request:
#### 203579d66ae912b7997b43e582ae078c2acfcb92
<pre>
[LBSE] Fix svg/clip-path/clip-path-on-overflowing.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=267832">https://bugs.webkit.org/show_bug.cgi?id=267832</a>

Reviewed by Nikolas Zimmermann.

Implement support for HTML elements referencing SVG clip paths
that rely on clipping through usage of a mask (support for pure
clip paths was already provided).

Since for non SVG we can&apos;t rely on objectBoundingBox(), add a
parameter to paintSVGClippingMask to pass the object bounding
box that should be used.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::paintClippingMask):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::paintSVGClippingMask const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):

Canonical link: <a href="https://commits.webkit.org/273495@main">https://commits.webkit.org/273495@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65ad0659f34c7e632f4a1489034269d63877a294

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14545 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32079 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36802 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11571 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36156 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10795 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10807 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31823 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39589 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32354 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32160 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36760 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10997 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8883 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34842 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12709 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8129 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11505 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1950,6 +1950,16 @@ void RenderBox::paintClippingMask(PaintInfo& paintInfo, const LayoutPoint& paint
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (document().settings().layerBasedSVGEngineEnabled() && style().clipPath() && style().clipPath()->type() == PathOperation::Reference) {
+        ASSERT(hasLayer());
```
I checked, this should be safe, since it is only used by RenderLayer.cpp.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.h:
```diff
@@ -101,6 +101,7 @@ class RenderLayerModelObject : public RenderElement {
     RenderSVGResourceMasker* svgMaskerResourceFromStyle() const;
 
+    void paintSVGClippingMask(PaintInfo&, const FloatRect& objectBoundingBox) const;
     void paintSVGClippingMask(PaintInfo&) const;
```
Do we need this accesor? Could get rid of it and only keep one.

rwlbuis replied:
Makes sense, we now need the FloatRect parameter.

nikolaszimmermann commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1950,6 +1950,16 @@ void RenderBox::paintClippingMask(PaintInfo& paintInfo, const LayoutPoint& paint
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (document().settings().layerBasedSVGEngineEnabled() && style().clipPath() && style().clipPath()->type() == PathOperation::Reference) {
+        ASSERT(hasLayer());
+        LayoutSize offsetFromRoot = layer()->offsetFromAncestor(layer()->root());
```
It's unusual to see layer tree information being used in RenderBox to drive clipping.
I'd investigate to avoid using the layer tree, to extract the information from the box model data - paintOffset, size(), location(), ...

rwlbuis replied:
Yes, it was old code, things work when using paintOffset and size() :)

nikolaszimmermann commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1951,6 +1951,14 @@ void RenderBox::paintClippingMask(PaintInfo& paintInfo, const LayoutPoint& paint
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (document().settings().layerBasedSVGEngineEnabled() && style().clipPath() && style().clipPath()->type() == PathOperation::Reference) {
+        LayoutRect paintRect = LayoutRect(paintOffset, size());
```
You can move that out.

rwlbuis replied:
Nice! Will do.

nikolaszimmermann reviewed (approved):
Good!

rwlbuis committed:
[LBSE] Fix svg/clip-path/clip-path-on-overflowing.html
https://bugs.webkit.org/show_bug.cgi?id=267832

Reviewed by Nikolas Zimmermann.

Implement support for HTML elements referencing SVG clip paths
that rely on clipping through usage of a mask (support for pure
clip paths was already provided).

Since for non SVG we can't rely on objectBoundingBox(), add a
parameter to paintSVGClippingMask to pass the object bounding
box that should be used.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::paintClippingMask):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::paintSVGClippingMask const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):

Canonical link: https://commits.webkit.org/273495@main
